### DISCUSSION TITLE
passing cookie with a request

### DISCUSSION THREAD

**Discussion by dsauna** [vercel/next.js#14473] [Help] [answered]
I have an external backend endpoint that relies on a cookie being present when calling. I tried adding the endpoint to getStaticProps but of course it fails and gives me a 401 because it doesn't have the cookie build time.

I can easily get it to work on the client side calling useSWR there and it works perfectly now my question is.

A) is there any way to get this working on getStaticProps? (please remember the cookie is generated by the backend.)

B) I could try and do it with getServerSideProps since that would have access to the cookie I believe. Would I gain anything by doing that?

C) I'm a newbie when it comes NextJs (did many tutorials and a couple of project but still missing a couple concepts) Is there any drawback of using useSWR directly and not using getServerSideProps or getStaticProps?

`const Component = () => {
  const dispatch = useDispatch();

  const fetcher = (request) =>
    axios(request).then((people) => {
      dispatch(setPeople(people.data));
      return people.data;
    });
  const { data, error } = useSWR('/rest/myEndpoint', fetcher);

  if (error) {
    console.log(error);
    return <Login />;
  }

  if (!data) return null;
  return <div/> //my component that work is here
}`

This is what I have so far

**Comment by surferwat** [comment, marked as answer]
Can you clarify what you mean by "it doesn't have the cookie at build time?"

The choice between getServerSideProps and getStaticProps largely depends on whether you need to fetch data at request time. Does the data you use in the page change with each request?

**Reply by dsauna** [reply]
Thanks @surferwat when we log in, we use react-google-login. When that returns. I call an external backend endpoint that generates a cookie. So Later when I call my endpoint to get my list of people, the endpoint returns a 401 if the cookie is not present with the endpoint. 

The data doesn't change with each request. It's fairly consistent. (It may get updated a couple of times a week with new people)

**Reply by surferwat** [reply]
If you decide to use getStaticProps, then you will need to rebuild your app every time that you want to reflect the updated data (i.e., the new people). Otherwise, you would be showing stale data until the next time you rebuild. If you want to be showing updated data, then getServerSideProps would better suit your needs. 

That said, you could access the cookie within the getServerSideProps block like so:

```
import Cookies from 'universal-cookie'

...

export async function getServerSideProps() {
  const cookies = new Cookies(context.req ? context.req.headers.cookie : null)
  const token = cookies.get('token')
  
  ...

}
```

Alternatively: https://flaviocopes.com/nextjs-cookies/

**Reply by dsauna** [reply]
Got it thanks! So given my example, I can't use getStaticProps, because I can't access cookies with getStaticProps, right? 

Is there any benefit to using the data getServerSideProps vs on the client the way I do it now?

**Comment by surferwat** [comment]
No problem. If the cookie is available at build time, you should be able to access the cookie using the same setup that I described previously for getServerSideProps.

You should use getServerSideProps if you need to pre-render the data. Otherwise it seems the recommended approach is to consider fetching data client side, the way you do it. 

reference: https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering

**Reply by dsauna** [reply]
Got it! So in general the only reason I would use getServerSideProps over client side rendering is if I need SEO, right? Because static html generation + client side data rendering should be more performant than having to rerender the entire page everytime with SSR, or am I getting things, wrong?

**Comment by rafaelalmeidatk** [comment]
> is there any way to get this working on getStaticProps?

When you generate a page with `getStaticProps` at build time, you are creating a content that all the users will see equally since you have no information about who is going to access this page. If you want to show dynamic content based on the auth state of the user, fetching it on the client-side is the recommended method. Be it with SWR, react-query, axios or any of these libraries.

> I could try and do it with getServerSideProps since that would have access to the cookie I believe. Would I gain anything by doing that?

Yeah, you will have access to the cookie since the server is not generating the page at build time. Each user will do a request to the page, and the server can render different page contents for each user using their information.

The drawback of this approach is that it is way slower than using `getStaticProps`, since you have to render the same page every time for each page request.

> I'm a newbie when it comes NextJs (did many tutorials and a couple of project but still missing a couple concepts) Is there any drawback of using useSWR directly and not using getServerSideProps or getStaticProps?

If you do not use `getServerSideProps` (or `getInitialProps`) then Next will automatically optimize the page for you, see: https://nextjs.org/docs/advanced-features/automatic-static-optimization

If you should or should not use the data-fetching methods depends on your needs. Keep in mind that `useSWR` will run only after you land on the page, if you inspect the page source the data won't be there. So if you need stuff like meta tags for social media, then you can't use only client-side data-fetching since the meta tags won't be ready when requested, you have to use some of the data-fetching methods, like `getStaticProps`

**Reply by dsauna** [reply]
Thank you much. I'm still failing to understand why would anyone ever use getServerSideProps over rendering data client side other than needing SEO

**Reply by rafaelalmeidatk** [reply]
SEO is not that much of a requirement to use `getServerSideProps` because nowadays the popular crawlers can still index your app even if you use client-side data-fetching, and you can use `getStaticProps` instead to avoid the higher TTFB (Time to first byte).

Some reasons that I can think of to use `getServerSideProps`:

- you have fast changing data and need to prerender it
- you need to prerender different content based on the request that can't be determined at build time
- for some reason, you need to use [Runtime Configuration](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration)

**Reply by dsauna** [reply]
Thank you so much!
